src/lib/schema/Schema.ts

```diff
@@ -0,0 +1,246 @@
+import { isFunction } from '@klasa/utils';
+import { SettingsFolder } from '../settings/SettingsFolder';
+import { fromEntries } from '../polyfills';
+
+/* eslint-disable no-dupe-class-members */
+
+export class Schema extends Map<string, SchemaFolder | SchemaEntry> {
+
+	/**
+	 * The base path for this schema.
+	 */
+	public readonly path: string;
+
+	/**
+	 * The type of this schema.
+	 */
+	public readonly type: 'Folder';
+
+	/**
+	 * The defaults for this schema.
+	 */
+	public readonly defaults: SettingsFolder;
+
+	/**
+	 * Whether or not this instance is ready.
+	 */
+	public ready: boolean;
+
+	/**
+	 * Constructs the schema
+	 */
+	public constructor(basePath = '') {
+		super();
+
+		this.ready = false;
+		this.path = basePath;
+		this.type = 'Folder';
+		this.defaults = new SettingsFolder(this);
+	}
+
+	/**
+	 * Adds or replaces an entry to this instance.
+	 * @param key The key of the entry to add
+	 * @param value The entry to add
+	 */
+	public set(key: string, value: SchemaFolder | SchemaEntry): this {
+		if (this.ready) throw new Error('Cannot modify the schema after being initialized.');
```

Comment:
```suggestion
		if (this.ready) throw 'Cannot modify the schema after being initialized.';
```

Wouldn't this suffice?

Reply:
No. This is not an error for potential users, but for developers.